Tests/NIOSSLTests/TLSConfigurationTest.swift

```diff
@@ -1302,6 +1322,182 @@ class TLSConfigurationTest: XCTestCase {
+                                          .TLS_PSK_WITH_AES_256_CBC_SHA]
+        
+        var serverConfig = TLSConfiguration.makePreSharedKeyConfiguration()
+        serverConfig.minimumTLSVersion = .tlsv1
+        serverConfig.maximumTLSVersion = .tlsv12
+        serverConfig.pskServerProvider = pskServerProvider
+        serverConfig.pskHint = "pskHint"
+        serverConfig.cipherSuiteValues = [.TLS_ECDHE_PSK_WITH_AES_256_CBC_SHA,
+                                          .TLS_ECDHE_PSK_WITH_AES_256_CBC_SHA,
+                                          .TLS_PSK_WITH_AES_128_CBC_SHA,
+                                          .TLS_PSK_WITH_AES_256_CBC_SHA]
+        try assertHandshakeSucceeded(withClientConfig: clientConfig, andServerConfig: serverConfig)
+    }
+    
+    func testTLSPSKFailure() throws {
+        // This test ensures that different PSKs used on the client and server fail when passed in.
+        let pskClientProvider: NIOPSKClientIdentityProvider = { (context: PSKClientContext) -> PSKClientIdentityResponse in
+            // Evaluate hint and clientIdentity to send back proper  PSK.
+            var psk = NIOSSLSecureBytes()
+            psk.append("hello".utf8)
+            return PSKClientIdentityResponse(key: psk, identity: "world")
+        }
+        
+        let pskServerProvider: NIOPSKServerIdentityProvider = { (context: PSKServerContext) -> PSKServerIdentityResponse in
+            // Evaluate hint and clientIdentity to send back proper  PSK.
+            var psk = NIOSSLSecureBytes()
+            psk.append("server".utf8) // Failure
+            return PSKServerIdentityResponse(key: psk)
+        }
+        
+        var clientConfig = TLSConfiguration.makeClientConfiguration()
+        clientConfig.certificateVerification = .none
+        clientConfig.minimumTLSVersion = .tlsv1
+        clientConfig.maximumTLSVersion = .tlsv12
+        clientConfig.pskClientProvider = pskClientProvider
+        clientConfig.pskHint = "pskHint"
+        
+        var serverConfig = TLSConfiguration.makePreSharedKeyConfiguration()
+        serverConfig.minimumTLSVersion = .tlsv1
+        serverConfig.maximumTLSVersion = .tlsv12
+        serverConfig.pskServerProvider = pskServerProvider
+        serverConfig.pskHint = "pskHint"
+        try assertHandshakeError(withClientConfig: clientConfig, andServerConfig: serverConfig, errorTextContainsAnyOf: ["SSLV3_ALERT_BAD_RECORD_MAC"])
+    }
+    
+    func testTLSPSKNoServerHint() throws {
+        let expectation = expectation(description: "pskClientProvider is called")
+        // This test ensures that different PSKs used on the client and server fail when passed in.
+        let pskClientProvider: NIOPSKClientIdentityProvider = { (context: PSKClientContext) -> PSKClientIdentityResponse in
+            expectation.fulfill()
+            // Ensure server hint is nil
+            XCTAssertEqual(context.hint, nil)
```

Comment:
We need to add equivalent checks on this in the other tests. We should probably also validate the identity and hint in the server callbacks in all tests.